Stable sort via bit magic

kentslaney opened the pull request:
I rolled back the `match_sort` implementation because of the `AndOp`/`OrOp` thing. Not quite sure where that kind of `SelectOp` is used?

kentslaney committed:
16 bit operand mask for stable argsort's bitcast

kentslaney committed:
up to 8.4e6 indices (untested)

kentslaney committed:
revert multikey sort matching

kentslaney committed:
multi-field floats move to the opposite outsides

kentslaney committed:
MIL `gather_along_axis` apparently casts the operand to 16 bits

kentslaney committed:
move bit shift to after gather

kentslaney committed:
fix gather's casted sign bit

kentslaney committed:
MIL op weirdness

First time trying antigravity; I'm so glad I didn't have to figure that one out myself

kentslaney committed:
more gather bitcasting problems

kentslaney committed:
bullet-proof shoes to remedy shooting self in foot

kentslaney committed:
nan, inf, and subnormals

kentslaney committed:
flake8 lint

kentslaney committed:
flake8 config

kentslaney committed:
fix 3 wide float32s

kasper0406 commented on stablehlo_coreml/converter.py:
```diff
@@ -418,21 +416,27 @@ def op_sort(self, context: TranslationContext, op: SortOp):
             raise ValueError("Unrecognized comparator format")
 
+        def precast_gather(tensor, indices):
+            if tensor.dtype == types.int32:
```
Fine workaround. Fysa, I think this is caused by a bug in coremltools that should be fixed by: https://github.com/apple/coremltools/pull/2626

kentslaney replied:
Oh, I actually rely on that bug for gathering the full 16 bits in stable sort (the part where I add `0x1_0000` to negative numbers). I would be able to transfer the same amount of info if they moved the sign bit over instead, but I've been avoiding that bit of int32s (thus the shift by `width - 1 - mask`). It's maybe possible to make it behave the same either way, but haven't thought it through yet.

I can add some of this context as comments too, I'm just used to and tired from working alone.

kasper0406 commented on stablehlo_coreml/sort_utils.py:
```diff
@@ -1,212 +1,227 @@
+    # Detect special values
+    # mb.is_nan/is_inf are not available, use comparisons
+    is_nan = mb.not_equal(x=x, y=x)
+    is_inf = mb.equal(x=mb.abs(x=x), y=width(float('inf')))
```
nit: I think you can use `np.inf` instead of `float('inf')` which looks slightly cleaner

kasper0406 reviewed (commented):
Good work getting this to work - some very impressive and complicated code. I won't really pretend to understand how it works, and I've added some comments where my understanding struggles the most.

kentslaney commented on stablehlo_coreml/converter.py:
```diff
@@ -418,21 +416,27 @@ def op_sort(self, context: TranslationContext, op: SortOp):
         # Apply the sort
         sort_dim, (key, ascending) = op.dimension.value, sort_keys[-1]
-        indices = mb.argsort(x=key, axis=sort_dim, ascending=ascending)
+        indices = stable_argsort(x=key, axis=sort_dim, ascending=ascending)
```
Oh yeah, I missed that it's less relevant for a single key sort. I think some of your code also had a check for HLO SortOp's `is_stable`, but I'm not sure if that's just permanently true in practice.

kentslaney commented:
>```python
> def test_argsort():
>     def sort_dim_0(k1):
>         k2 = jnp.arange(k1.shape[0], dtype=jnp.int32)
>         return jax.lax.sort([k1, k2], dimension=0, num_keys=2)
>
>     run_and_compare_specific_input(sort_dim_0, (
>         jnp.array([10, 5, 10, 0x1_0010, 10, 0x5_0000, 10, 5, 10, 0x1_0000, 5], dtype=jnp.int32),
>     ))
>```
> ```python
> E            ACTUAL: array([ 0,  0,  5,  5,  5, 10, 10, 10, 10, 10, 16], dtype=int32)
> E            DESIRED: array([     5,      5,      5,     10,     10,     10,     10,     10,
> E                   65536,  65552, 327680], dtype=int32)
> ```

@kasper0406 This is actually because the gather of `k1` from sorting `k2` overflows, which I don't think I can fix (besides in the trivial case where there are no changes). In the original it's `argsort` instead of `sort`, so you can tell it's stable because the elements have been moved the minimum amount possible.

kasper0406 commented on stablehlo_coreml/sort_utils.py:
```diff
@@ -1,212 +1,227 @@
+from coremltools.converters.mil.mil import types
+
+
+def bitcast_fp(x):
```
Right. I just have a hard time convincing myself if it is one of those things which is always doing the correct thing, or doing the correct thing "most of the time"

kentslaney committed:
large sorts test

kentslaney commented:
I can't seem to independently reproduce the overflow behavior that I thought was causing the problem

```python
import numpy as np
import coremltools as ct
from coremltools.converters.mil.mil import Builder as mb
from coremltools.converters.mil.mil import types

def check_size(n, target):
    @mb.program(input_specs=[
        mb.TensorSpec(shape=(n,), dtype=types.int32),
        mb.TensorSpec(shape=(1,), dtype=types.int32)
    ])
    def prog(x, indices):
        out = mb.gather_along_axis(x=x, indices=indices, axis=0, name="output")
        return out

    model = ct.convert(prog, source='milinternal', convert_to="mlprogram", compute_precision=ct.precision.FLOAT32)

    x_val = np.arange(n, dtype=np.int32)
    indices_val = np.array([target], dtype=np.int32)

    prediction = model.predict({"x": x_val, "indices": indices_val})
    result = prediction["output"]

    print(f"{n:#_x} {target:#_x} {result[0]:#_x}")

check_size(0x10_0010, 0x10_0008)
```

kentslaney commented:
Issue description starting from the [`sort-gather-overflow` branch](https://github.com/kentslaney/stablehlo-coreml/tree/f7070c72be3a6a575158c9a2b52bd599bcc44d03)

`env/bin/python tests/mb_gather_overflow.py` suggests that `mb.gather_along_axis` shouldn't have a problem with operands that don't fit in 16 bits. Why, then, does `env/bin/pytest tests -k test_gather_cast` seem to wrap when the gathered input is larger than 16 bits?

Interestingly, switching the first `stable_argsort` to `mb.argsort` in `op_sort` removes the wrapping issue (though introduces an order issue) even though the second input with the wrapping problem only goes through the final `mb.gather_along_axis` call and the behavior persists with an explicit cast of indices to int32, which should remove any discrepancies in input types.

kentslaney commented:
I've yet to check whether patching to apple/coremltools#2626 makes a difference (dumping note-to-self progress updates here because of an intermission)

kentslaney committed:
remove tests that overflow `common::add_int16_cast`

kentslaney committed:
remove pre-cast gather to support pipelines without the int16 cast

tests pass with build from git+https://github.com/apple/coremltools.git@refs/pull/2626/head

kentslaney committed:
remove int16 casting pass from test pipeline

kentslaney committed:
total sort specifics and unstable sort shortcut

kentslaney committed:
error message for bitcast int's width

kentslaney committed:
move assert to branch

besides consistency, it makes code coverage reports simpler

kentslaney committed:
nan ordering test (fails)

kentslaney committed:
compressed nan ordering test

kentslaney committed:
Merge branch 'main' into sort-bitcast

kentslaney committed:
Merge branch 'main' into sort-bitcast

kentslaney committed:
fix problem merging from main

I should really stop using the web interface to resolve merge conflicts, the gh cli is probably worth getting used to anyway

kentslaney committed:
test coverage (`test_sort_int16_casting` now broken on Apple Silicon)

kentslaney committed:
test descending stable sort

kentslaney committed:
dtype width utility function

kentslaney committed:
tests bitcast workaround for fp16

kentslaney committed:
Inf < NaN

kentslaney commented:
I'm getting [inconsistent results](https://github.com/kentslaney/stablehlo-coreml/blob/e45274323f14b316a0b167f477223d21851b5cf1/consistency.sh#L3-L13) and sometimes a segfault on what I expect to be a deterministic pipeline.

kentslaney commented:
>Btw. if you have Discord / other chat client, I'm happy to iterate there as well, if easier.

My Discord is @1i1iii1 ([relevant](https://xkcd.com/1105))

kentslaney committed:
MIL-based reduction initialization to support `skip_const_by_size`

kentslaney committed:
rank 0 tiling bug

kentslaney committed:
use `mb.range_1d` for `hlo.iota` instead of `np.arange`

kentslaney committed:
move iota into utility function

kentslaney committed:
negative axis support for iota

kentslaney committed:
flake8 lint

kentslaney committed:
Merge branch 'reduction-init' into sort-bitcast
